### PR TITLE
More changes toward an end-to-end flow with the new APIS

### PR DESCRIPTION
**Pull request by chavdar**
Changes to the new API:
- Made GobblinInstanceLauncher extend configurable since it is responsible of holding the system configuration
- Added initial version of JobExecutionDriver and modified JobExecutionLauncher to return JobExecutionDriver
- Added a JobLauncherExecutionDriver which acts as an adapter to the legacy JobLauncher/JobLauncherFactory API
- Reworked JobSpec.Builder to be convention-based and be able to reuse some of the properties in the Job configuration
- Added StandardGobblinInstanceDriver and StandardGobblinInstanceLauncher-- a very rudimentary implementations of GobblinInstanceDriver and GobblinInstanceLauncher

Changes to legacy code:
- Documentation and helper methods in JobState to decrease reliance on direct access to job properties
- Helper methods in JobContext
- Added a method JobLauncherFactory. newJobLauncher() method to allow instantiating a specific JobLauncher

@ibuenros @autumnust can you review ?

### CODE REVIEW

File: gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java [hunk 1]
```diff
@@ -36,4 +54,86 @@ public JobExecutionState(JobSpec jobSpec, JobExecution jobExecution) {
...
+      case SUCCESSFUL: this.switchToSuccessful(); break;
+      case FAILED: this.switchToFailed(); break;
+      case COMMITTED: this.switchToCommitted(); break;
+      case CANCELLED: this.switchToCancelled(); break;
+      default: throw new Error("Unexpected state");
+    }
+  }
+
+  public void switchToPending() {
+    Preconditions.checkState(null == this.runningState, "unexpected state:" + this.runningState);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java]: nit: `String.format("unexpected stage transition: %s to %s", this.runningState, PENDING)`, also for the other similar methods.

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java [hunk 2]
```diff
@@ -137,7 +137,8 @@ public JobSpecRunnable(JobSpec jobSpec) {
 
     @Override
     public void run() {
-       _jobLauncher.launchJob(_jobSpec, new JobExecutionListener());
+       _jobLauncher.launchJob(_jobSpec);
+       // new JobExecutionListener();
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java]: This is incomplete and will get changed somewhat in a following commit;

File: gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java (hunk 1 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java]: Will fix.

**Comment by chavdar** [comment]
Since a lot of changes have accumulated, I'll re-summarize the changes:

Changes to the new API:

* GobblinInstanceLauncher/GobblinInstanceDriver
    - Renamed GobblinInstanceLauncher.getName() to GobblinInstanceLauncher.getInstanceName()
    - Made GobblinInstanceLauncher extend configurable since it is responsible of holding the system configuration

* JobCatalog/JobSpec:
    - Refactored the register/unregister methods into a separate interface - JobCatalogListenersContainer
    - Reworked JobSpec.Builder to be convention-based and be able to reuse some of the properties in the Job configuration

*  JobExecutionLauncher/JobExecutionDriver
    - Added the class JobExecutionDriver to control/monitor the execution of a single gobblin job
    - JobExecutionLauncher is not a factory for JobExecutionDriver instances
    - Add JobExecutionStateListenerContainer interface for managing JobExecutionStateListener instances

* JobExecutionStatus/JobExecutionState
    - JobExecutionStatus is now an interface
    - JobExecutionState is now responsible for the state management (including callbacks)

Implementations:

- Added a sample implementation JobLauncherExecutionDriver of JobExecutionDriver which acts as an adapter to the legacy JobLauncher/JobLauncherFactory API
- Added DefaultGobblinInstanceDriverImpl which accepts all components but it is not responsible for managing them. This implementation can be used as a shell to implement a typical Gobblin job lifecycle
- Added StandardGobblinInstanceDriver which will be responsible for instantiating and managing the instance components and passing them to DefaultGobblinInstanceDriverImpl
- Added StandardGobblinInstanceLauncher-- a very rudimentary implementation to instantiate StandardGobblinInstanceDriver
- Standard implementations for JobCatalogListenersContainer (JobCatalogListenersList) and JobExecutionStateListenerContainer (JobExecutionStateListeners)
- The above use a generic callback infrastructure in gobblin-utility/gobblin/util/callbacks
  - Callbacks a typically represented through a Function<L, R> where L is the listener type and R is the callback result (or Void)
  - There is a helper class Callback<L, R> to let adding some human-readable description of the callback
  - Callbacks are created by a CallbackFactory which is responsble for binding any parameters that need to be passed to the callback
  - Callbacks are managed by CallbacksDispatcher

Changes to legacy code:

- Documentation and helper methods in JobState to decrease reliance on direct access to job properties
- Added a description of current state diagram for RunningState based on code inspection
- Helper methods in JobContext
- Added a method JobLauncherFactory. newJobLauncher() method to allow instantiating a specific JobLauncher type instead of relying on the job config

**Comment by chavdar** [comment]
As discussed, there some follow-up items:
- Have an internal implementation of JobExecutionStatus in JobExecutionState that can be easily serialized/deserialized
- Add more javadocs to CallbacksDispatcher to explain usage

I'll make the above in a separate PR since my local copy is already ahead of this PR and I'd rather don't merge any more changes in this PR.

File: gobblin-utility/src/main/java/gobblin/util/callbacks/CallbacksDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  public <R> CallbackResults<L, R> execCallbacks(CallbackFactory<L, R> callbackFactory)
+         throws InterruptedException {
+    Preconditions.checkNotNull(callbackFactory);
+    List<L> listeners = getListeners();
+
+    return execCallbacks(callbackFactory, listeners);
+  }
+
+  private <R> CallbackResults<L, R> execCallbacks(CallbackFactory<L, R> callbackFactory, List<L> listeners)
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/callbacks/CallbacksDispatcher.java]: Couldn't you pass the `Callback` directly?

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java [hunk 4]
```diff
@@ -137,7 +137,8 @@ public JobSpecRunnable(JobSpec jobSpec) {
 
     @Override
     public void run() {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java]: Run is generally not used in an `AbstractIdleService`, should you use a `AbstractExecutionThreadService` instead?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java [hunk 5]
```diff
@@ -1,154 +1,148 @@
...
     }
   }
 
-  public void callbackOneListener(Callback callback, JobCatalogListener listener) {
-    String callbackMsg = null;
-    if (_debugLogEnabled) {
-      callbackMsg = "callback " + callback + " on " + listener;
-      _log.debug("Started: " + callbackMsg);
-    }
+  public void callbackOneListener(Function<JobCatalogListener, Void> callback,
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java]: What does this method do?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java [hunk 6]
```diff
@@ -1,154 +1,148 @@
...
-      jobSpecToParamString(param1, sb);
-      if (null != param2) {
-        sb.append(']');
-        sb.append(',');
-        jobSpecToParamString(param2, sb);
-      }
-      sb.append(']');
-      sb.append(')');
-      return sb.toString();
+  public static class AddJobCallback extends Callback<JobCatalogListener, Void> {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java]: Could this be inner static classes of `JobCatalogListener` instead? They are not specific to this class.

File: gobblin-utility/src/main/java/gobblin/util/callbacks/CallbacksDispatcher.java (hunk 3 above)
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/callbacks/CallbacksDispatcher.java]: Yeah, fixed.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        case FAILED:
+        case COMMITTED:
+        case CANCELLED: {
+          // Nothing to do
+          break;
+        }
+      }
+    }
+
+    _legacyLauncher.close();
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: Will this prevent the job from running?

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java (hunk 5 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java]: This is needed for the contract, that when we add a new listener, it should get callbacks for all existing jobs. See InMemoryJobCatalog.addListener() .

File: gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java (hunk 6 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java]: Sure. Makes sense.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    public MyJobListener() {
+      super(Optional.of(JobLauncherExecutionDriver.this._log));
+    }
+
+    @Override
+    public void onJobPrepare(JobContext jobContext) throws Exception {
+      super.onJobPrepare(jobContext);
+      _jobContext = jobContext;
+      _jobState.switchToPending();;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: extra `;`

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+          // Nothing to do
+          break;
+        }
+      }
+    }
+
+    _legacyLauncher.close();
+  }
+
+  class MyJobListener extends AbstractJobListener {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: Can you change the name of the class? Maybe `JobListenerToJobStateBridge` or something more meaningful?

File: gobblin-runtime/src/test/java/gobblin/runtime/std/TestJobExecutionStateListeners.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import gobblin.runtime.api.JobExecutionStateListener;
+import gobblin.runtime.api.JobSpec;
+
+/**
+ * Unit tests for {@link JobExecutionStateListeners}.
+ */
+public class TestJobExecutionStateListeners {
+
+  @Test
+  public void testHappyPath() {
```
Comment (ibuenros) [inline gobblin-runtime/src/test/java/gobblin/runtime/std/TestJobExecutionStateListeners.java]: Interesting name :P

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java (hunk 9 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: Done.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java (hunk 8 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: Fixed.

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java (hunk 4 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java]: This is actually a method of JobSpecRunnable which gets passed to the JobExecutionScheduler.

File: gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java (hunk 7 above)
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java]: Yes, this shut downs the ExecutorService in AbstractJobLauncher.

**Pull request merged**